### PR TITLE
utils: Fix broken Windows filelock

### PR DESCRIPTION
**Pull request by ken2812221**
Fix broken filelock on Windows, also add a test for this. It's a regression introduced by #13862.

### CODE REVIEW

**Comment by ken2812221** [comment]
For reference, the boost implementation is in https://github.com/boostorg/interprocess/blob/8b3621353017aa527a22124655c9458d0b64b358/include/boost/interprocess/detail/os_file_functions.hpp#L230-L243

**Comment by donaloconnor** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14426/commits/6d336e39dffb4951f09beedc478ba425c38c60b2

File: src/fs.cpp [hunk 1]
```diff
@@ -89,7 +89,7 @@ bool FileLock::TryLock()
         return false;
     }
     _OVERLAPPED overlapped = {0};
-    if (!LockFileEx(hFile, LOCKFILE_EXCLUSIVE_LOCK | LOCKFILE_FAIL_IMMEDIATELY, 0, 0, 0, &overlapped)) {
+    if (!LockFileEx(hFile, LOCKFILE_EXCLUSIVE_LOCK | LOCKFILE_FAIL_IMMEDIATELY, 0, -1, -1, &overlapped)) {
```
Comment (Empact) [inline src/fs.cpp]: nit: `std::numeric_limits<DWORD>::max()` or explicitly casting to unsigned might be more clear.

**Comment by donaloconnor** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14426/commits/474d022036ca2114750fc650f6712a99eda1bad4

File: test/functional/feature_filelock.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+from test_framework.util import assert_equal
+
+import os
+
+class FilelockTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+
+    def get_node_output(self, *, index, ret_code_expected):
```
Comment (promag) [inline test/functional/feature_filelock.py]: Move to node class as `get_output`? If that's controversial then extract to a module function for now? Then below:

```py
_, output = get_node_output(node=self.nodes[1], ret_code_expected=1)
```

Comment (ken2812221) [reply test/functional/feature_filelock.py]: Done

File: test/functional/feature_filelock.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    return out, err
+
+class FilelockTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+
+    def run_test(self):
+        self.nodes[1].stop()
+        self.log.info("Using datadir " + self.nodes[0].datadir)
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply test/functional/feature_filelock.py]: Done

**Comment by jnewbery** [comment]
This seems like a reasonable change (although I know nothing about the windows API). `LockFileEx` documentation is here: https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-lockfileex.

The test case is good, but it has the wrong file permissions and replicates functionality that is already available in the test framework. I've reimplemented it here: https://github.com/jnewbery/bitcoin/tree/pr14426.1 which uses the `assert_start_raises_init_error()` method. It's more concise, runs faster, also covers trying to open a wallet being used by a different bitcoind process, and I think is clearer. Feel free to take that commit in this PR.

**Comment by ken2812221** [comment]
Thanks, I didn't know test framework that much, just trying to do what I want. Your commit looks good to me, I'll take it. The only thing I change is to skip wallet test if it's not compiled.

**Review by ryanofsky — APPROVED** [review]
utACK c7296a98e56a3a38730575c7aa7f524f10e2a0a9, which seems to be updated with John's test.

Although the [LockFileEx](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-lockfileex) documentation just says:

> Locking a region that goes beyond the current end-of-file position is not an error

The [LockFile](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-lockfile) documentation is more specific and says:

> You can lock bytes that are beyond the end of the current file. This is useful to coordinate adding records to the end of a file

which I think explains why the fix works.

**Comment by jnewbery** [comment]
utACK c7296a98e56a3a38730575c7aa7f524f10e2a0a9.

I think the `feature_notifications.py` failure in appveyer is unrelated. The `feature_filelock.py` failure is due to a path slash mismatch: `C:/Users\appveyor` instead of `C:\Users\appveyor`.

**Comment by donaloconnor** [comment]
@ken2812221 should we explicitly call UnlockFile before closing the handle. The OS does unlock it but it's not instant potentially. Maybe safer to unlock before closing

**Comment by ken2812221** [comment]
@donaloconnor I am not sure what condition would be unsafe if we don't unlock the file.

**Comment by donaloconnor** [comment]
@ken2812221 I did a test there across 2 processes to see if there's any weird race conditions with not unlocking before  close and there's not. I thought being explicit might be better. Anyway no big deal.

tested ACK.

**Comment by jnewbery** [comment]
utACK 369244f654c9e36b803e841eb30fd0aa2960087a

**Review by ryanofsky — COMMENTED** [review]
utACK 369244f654c9e36b803e841eb30fd0aa2960087a, only change is setting ErrorMatch.PARTIAL_REGEX flag in test.

**Comment by sipa** [comment]
utACK 369244f654c9e36b803e841eb30fd0aa2960087a

**Pull request merged**